### ISSUE TITLE
apt packages are version 1.12.0-rc.1-00 now -> fails deploying

### ISSUE THREAD

**Issue by sebastian-su** [geerlingguy/ansible-role-kubernetes#18]
Hi

obviously yesterday new rc versions for kubernetes 1.12 got uploaded to the official apt repo. That breaks installing from your module. I tried some approaches but I am not that fluent in ansible yet.

somehow we need to make ansible call "apt-get install kubectl=v*** kubeadm=v*** ... "
All options I found in ansible where you can specify a version will lead to upgrade & downgrade behavior. Maybe you have an idea ;)

my quick fix is deploying a /etc/apt/preferences.d/kubectl file with 
```
Package: kubectl
Pin: version 1.11.*
Pin-Priority: 1000

Package: kubeadm
Pin: version 1.11.*
Pin-Priority: 1000

Package: kubelet
Pin: version 1.11.*
Pin-Priority: 1000
```

It needs to run before the include or your role of course.

let me know if i can help.

**Comment by rhino5oh** [comment]
I ran into this same issue. I put the workaround below.  Obviously you can change the set_fact to whatever version you want.

**Also, very important note:  Order of the packages matters!!** Why?  Because the 'kubeadm' package will automatically install 'kubectl'.  BUT if you say kubeadm=11.1.0-00, it will NOT install the same version of kubectl.  It will install the latest.  At the time of writing, that is 1.12.0-rc.1.

So, what happens is, kubeadm 11.1.0-00 gets installed.  It automatically installs kubectl 1.12.0-rc.1, then, if you have kubectl AFTER kubeadm, geerlingguy's role trys to install kubectl 11.1.0-00, which is a downgrade, and you get the following error:   "Packages were downgraded and -y was used without --allow-downgrades".  So, use the order I show below and you should be OK for now



      - set_fact:
          kube_version: "-1.11.1-0"
          cni_version: "-0.6.0-0"
        when: ansible_os_family == 'RedHat'

      - set_fact:
          kube_version: "=1.11.1-00"
          cni_version: "=0.6.0-00"
        when: ansible_os_family == 'Debian'

      - name: "Running geerlingguy.kubernetes role"
        include_role:
          name: geerlingguy.kubernetes
        vars:
          kubernetes_packages:
            - name: kubelet{{ kube_version }}
              state: present
            - name: kubectl{{ kube_version }}
              state: present
            - name: kubeadm{{ kube_version }}
              state: present
            - name: kubernetes-cni{{ cni_version }}
              state: present

**Comment by geerlingguy** [comment]
Gah... sorry about this, I'll hopefully have something set up so this can be a bit more easily managed. Haven't had to deal with projects releasing non-stable releases into their apt / rpm repos much before.

**Comment by geerlingguy** [comment]
See: https://github.com/coreos/flannel/issues/1044

**Comment by geerlingguy** [comment]
For Debian, I'm considering adding an apt preferences file as @sebastian-su suggests in his original comment. For RedHat, I might need to use yum-versionlock, still looking into how to lock it into a major release (e.g. 1.10 or 1.11).

**Comment by geerlingguy** [comment]
Yay on RHEL at least, initial testing seems to be working okay:

```
# kubelet --version
Kubernetes v1.11.3
```

**Comment by geerlingguy** [comment]
Still working on Debian/Ubuntu apt pinning.

**Comment by geerlingguy** [comment]
Fixed in the `2.0.0` release. Note the new major version, since this required a breaking change:

  * `kubernetes_version` is now used to drive the `kubeadm init` version (new var `kubernetes_version_kubeadm`) and Debian/Ubuntu apt version pinning preferences.
  * `kubernetes_version_rhel_package` is new, and drives a (for now, non-overridable) package list specific to RedHat/CentOS in `vars/RedHat.yml`.

But it seems a _lot_ more stable, especially since I also fixed #17 alongside this to get Flannel working correctly out of the box.

I also added a few more checks and switched tests over to Molecule so hopefully things get better and more stable (just like they—slowly—are in the wider Kubernetes ecosystem).
